### PR TITLE
WIP Chaos Bag

### PR DESCRIPTION
**Pull request by suxur**

### CODE REVIEW

File: actions/types.ts [hunk 1]
```diff
@@ -265,6 +271,7 @@ export interface Campaign {
   latestDeckIds?: number[]; // deprecated
   investigatorData: InvestigatorData;
   chaosBag: ChaosBag;
+  chaosBagResults: ChaosBagResults;
```
Comment (zzorba) [inline actions/types.ts]: So because this goes into redux (and is persisted on devices), we'll need to make it optional (`?: ChaosBagResults`) and handle giving it a default case in our code.

File: actions/types.ts [hunk 2]
```diff
@@ -182,6 +182,12 @@ export interface InvestigatorCampaignNoteCount {
   custom?: boolean;
 }
 
+export interface ChaosBagResults {
+  currentToken?: ChaosTokenType;
+  drawnTokens: ChaosTokenType[];
+  total: number;
```
Comment (zzorba) [inline actions/types.ts]: What's total -- the total # of tokens in the bag / drawn?  Or total available?  I might suggest making this derived state (based on what's drawn + sealed + available).

What behavior do we want to happen when a token is added/removed from the bag (probably nothing right, the sealed tokens stay sealed, drawn tokens stayed drawn, etc).

File: actions/types.ts [hunk 3]
```diff
@@ -182,6 +182,12 @@ export interface InvestigatorCampaignNoteCount {
   custom?: boolean;
 }
 
+export interface ChaosBagResults {
+  currentToken?: ChaosTokenType;
```
Comment (zzorba) [inline actions/types.ts]: How heavy is the divide between current + drawn.  Is there a world where current tokens is just the 'first' drawn token?

File: components/campaign/CampaignChaosBagView/index.tsx [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+
+interface State {
+    chaosBagResults: ChaosBagResults;
+    visible: boolean;
+    iconKey?: ChaosTokenType;
+}
+
+type Props = CampaignChaosBagProps & NavigationProps;
+
+// TODO: need to change it to pull chaosBag from redux state (pass the campaignId, and fetch campaign + get it from redux)
```
Comment (zzorba) [inline components/campaign/CampaignChaosBagView/index.tsx]: Happy to help with this, or I can walk you through the process. The `CampaignDrawWeaknessDialog` is probably a good example of one that updates a piece of a campaign state in a similar way to this.  Our redux store for campaigns is just done by updating a 'partial' campaign, so you probably don't have to touch the literal redux store (just map the state out + use the existing `updateCampaign` action)

File: components/campaign/CampaignChaosBagView/index.tsx [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+        drawnTokens: [],
+        total: 0,
+      },
+      visible: true,
+    };
+
+    this._navEventListener = Navigation.events().bindComponent(this);
+  }
+
+  componentDidMount() {
```
Comment (zzorba) [inline components/campaign/CampaignChaosBagView/index.tsx]: So the existing edit dialog did all this 'back button' override so there would be a confirmation when you press back.  I THINK this screen might be able to get away with the standard back functionality, since theoretically the set of drawn tokens + sealed tokens is 'non-destructive'?

What do you think?   If so, I think you can just remove these back button event listeners (as well as the handling of 'back' and 'androidBack') and just update the redux store on every draw/seal/clear?

File: components/campaign/CampaignChaosBagView/index.tsx [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+    if (buttonId === 'back' || buttonId === 'androidBack') {
+      this._handleBackPress();
+    } else if (buttonId === 'edit') {
+      this._showChaosBagDialog();
+    }
+  }
+
+  _handleDrawTokenPressed = () => {
+    this._drawToken();
+  }
```
Comment (zzorba) [inline components/campaign/CampaignChaosBagView/index.tsx]: semicolons on these bound functions.

File: components/campaign/CampaignChaosBagView/index.tsx [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+          weight_sum += item;
+          weight_sum = +weight_sum.toFixed(2);
+        }
+
+        if (random_num <= weight_sum) {
+          const tokenName = list[i];
+          let drawnToken = null;
+          CHAOS_TOKENS.forEach(token => {
+            if (token === tokenName) {
+              drawnToken = token;
```
Comment (zzorba) [inline components/campaign/CampaignChaosBagView/index.tsx]: You could just do the return here directly -- also I believe CHAOS_TOKENS can be indexed into (or if not there should be a map that can be nearby it).

File: components/campaign/CampaignChaosBagView/index.tsx [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+  _getRandomChaosToken = (chaosBag: ChaosBag) => {
+    const list = keys(chaosBag);
+    const weight = values(chaosBag);
+
+    const totalWeight = weight.reduce((prev, cur) => {
+      return (prev && cur) ? (prev + cur) : 0;
+    });
+
+    if (totalWeight) {
+      const random_num = this._getRandomNumber(0, totalWeight);
```
Comment (zzorba) [inline components/campaign/CampaignChaosBagView/index.tsx]: Given our set is always pretty small (14-20 tokens), it might be simpler to just make an array of all available tokens, `shuffle` them (using lodash) and take the first?  

If they draw another token, just repeat the above (with one less available now)?

I'm sure this method would work as well, it's just a bit harder to reason about the corner cases in it, and random things can be hard to test.

File: components/campaign/CampaignChaosBagView/index.tsx [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+      let weight_sum = 0;
+      const length = list.length;
+      for (let i = 0; i < length; i++) {
+        const item = weight[i];
+        if (item) {
+          weight_sum += item;
+          weight_sum = +weight_sum.toFixed(2);
```
Comment (zzorba) [inline components/campaign/CampaignChaosBagView/index.tsx]: Whats the `+` do here?  Not sure I'm familiar with that operator.

File: components/campaign/CampaignChaosBagView/index.tsx [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+          drawnTokens: chaosBagResults.drawnTokens,
+          total: chaosBagResults.total + 1,
+        },
+      });
+    } else {
+      this.setState({
+        chaosBagResults: {
+          currentToken: undefined,
+          drawnTokens: [],
+          total: chaosBagResults.total,
```
Comment (zzorba) [inline components/campaign/CampaignChaosBagView/index.tsx]: Got it, so total is just the all time count of how many tokens you've drawn?

File: components/campaign/CampaignChaosBagView/index.tsx [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+      const newIconKey = this._getRandomChaosToken(chaosBag);
+      this.setState({
+        chaosBagResults: {
+          currentToken: newIconKey,
+          drawnTokens: chaosBagResults.drawnTokens,
+          total: chaosBagResults.total + 1,
+        },
+      });
+    } else {
+      this.setState({
```
Comment (zzorba) [inline components/campaign/CampaignChaosBagView/index.tsx]: I might see if I could break this into two different functions, since they seem kind of unrelated?  Maybe I don't quite understand what's happening here -- so the if branch is if you have NO token drawn OR you are asking for a draw + refresh.  Else block is... just a refresh?

Comment (zzorba) [reply components/campaign/CampaignChaosBagView/index.tsx]: Seems like else block is basically clearTokens, maybe just update this to call that.

File: components/campaign/CampaignChaosBagView/index.tsx [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+
+    if (drawnTokens.length > 0) {
+      return drawnTokens.map(function(token, index) {
+        return (
+          <ChaosToken key={index} iconKey={token} small />
+        );
+      });
+    }
+
+    return <Text style={[styles.drawTokenText, typography.text]}>Tap token to draw</Text>;
```
Comment (zzorba) [inline components/campaign/CampaignChaosBagView/index.tsx]: Throw the copy string here in a `{t'Tap token to draw'}` only with backtick instead of '

File: components/campaign/CampaignChaosBagView/index.tsx [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+  }
+
+  _renderDrawButton = () => {
```
Comment (zzorba) [inline components/campaign/CampaignChaosBagView/index.tsx]: Since we don't usually pass the render functions, I just keep them normal most often.

```
renderDrawButton() {
}
```

File: components/campaign/ChaosToken.tsx [hunk 14]
```diff
@@ -0,0 +1,79 @@
+import React from 'react';
+import { ImageBackground, StyleSheet, Text, View } from 'react-native';
+import { ChaosTokenType } from '../../constants';
+import ChaosTokenIcon from '../core/ChaosTokenIcon';
+import { iconSizeScale } from '../../styles/space';
+import DeviceInfo from 'react-native-device-info';
```
Comment (zzorba) [inline components/campaign/ChaosToken.tsx]: If you don't mind, I like to keep the npm package imports at the top, with the local ones down below.  A few of the 'ttag' rules don't follow this because I did a global find replace, but I try to fix them when I come across them.

File: components/campaign/ChaosToken.tsx [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+
+export default class ChaosToken extends React.Component<Props> {
+
+  _renderIcon = () => {
+    const { iconKey, small } = this.props;
+    const size = small ? 25 : 50;
+
+    if (iconKey) {
+      return (
+        <Text>
```
Comment (zzorba) [inline components/campaign/ChaosToken.tsx]: Does it work without the <Text> wrapper, I think ChaosTokenIcon should automatically create a text/icon as needed.

File: components/campaign/ChaosToken.tsx [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+          <ChaosTokenIcon icon={iconKey} size={size * SCALE * iconSizeScale} color="#fff" fontFamily="Teutonic" />
+        </Text>
+      );
+    }
+  }
```
Comment (zzorba) [inline components/campaign/ChaosToken.tsx]: Semicolon at end of the bound function, I believe `npm run build` would complain about this.

File: actions/types.ts (hunk 2 above)
Comment (suxur) [reply actions/types.ts]: Total should be renamed to something like `totalTokensDrawn`. It really isn't even necessary, just some added flare.

When a token is added/removed from the bag, everything should continue to work just with the new state of the bag.

File: actions/types.ts (hunk 3 above)
Comment (suxur) [reply actions/types.ts]: Not really that heavy, it could easily be the first from the drawn tokens.

File: components/campaign/CampaignChaosBagView/index.tsx (hunk 5 above)
Comment (suxur) [reply components/campaign/CampaignChaosBagView/index.tsx]: Yeah it it's non-destructive so that should be fine as long as the state of the drawn and sealed tokens persist.

File: components/campaign/CampaignChaosBagView/index.tsx (hunk 9 above)
Comment (suxur) [reply components/campaign/CampaignChaosBagView/index.tsx]: I feel like this is a typo. Let me look into it.

File: components/campaign/CampaignChaosBagView/index.tsx (hunk 11 above)
Comment (suxur) [reply components/campaign/CampaignChaosBagView/index.tsx]: I actually refactored this already to a method named just that! `clearTokens()`

File: components/campaign/ChaosToken.tsx (hunk 15 above)
Comment (suxur) [reply components/campaign/ChaosToken.tsx]: I'll take a look

File: components/campaign/ChaosToken.tsx (hunk 16 above)
Comment (suxur) [reply components/campaign/ChaosToken.tsx]: WebStorm wasn't complaining based on eslint but I'll see what `npm run build` has to say

File: components/campaign/CampaignChaosBagView/index.tsx [hunk 17]
```diff
@@ -0,0 +1,363 @@
...
+};
+
+function mapStateToProps(
+  state: AppState,
+  props: NavigationProps & CampaignChaosBagProps
+): ReduxProps {
+  const campaign = getCampaign(state, props.campaignId);
+  return {
+    chaosBag: campaign ? campaign.chaosBag : {},
+    chaosBagResults: campaign ? campaign.chaosBagResults : EMPTY_CHAOS_BAG,
```
Comment (zzorba) [inline components/campaign/CampaignChaosBagView/index.tsx]: Make this `chaosBagResults: (campaign && campaign.chaosBagResults) || EMPTY_CHAOS_BAG,` so it handles old campaigns where chaosBagResults isn't set.

File: components/campaign/CampaignChaosBagView/index.tsx [hunk 18]
```diff
@@ -0,0 +1,363 @@
...
+  };
+
+  drawToken = () => {
+    const { campaignId, chaosBag, chaosBagResults } = this.props;
+
+    const newIconKey = this.getRandomChaosToken(chaosBag);
+    const drawnTokens = chaosBagResults.drawnTokens;
+
+    if (newIconKey) {
+      drawnTokens.push(newIconKey);
```
Comment (zzorba) [inline components/campaign/CampaignChaosBagView/index.tsx]: So this is the problem here.  Because the drawnTokens is being directly mutated (instead of making a copy of it), the prop isn't invalidating (since its the same object as before).  You always want to make a copy of a prop to modify, never edit them directly, or it can break the way that Redux tracks invalidation (basically if its the same 'shallow' object, it doesn't look any deeper).
Using a spread operator or a slice() before doing the push will fix this.
`const drawnTokens = [...chaosBagResults.drawnTokens];`

File: components/campaign/CampaignChaosBagView/index.tsx [hunk 19]
```diff
@@ -0,0 +1,363 @@
...
+  clearTokens = () => {
+    const { campaignId, chaosBagResults } = this.props;
+
+    const newChaosBagResults = {
+      drawnTokens: [],
+      sealedTokens: chaosBagResults.sealedTokens,
+      totalDrawnTokens: chaosBagResults.totalDrawnTokens,
+    };
+
+    updateCampaign(campaignId, { chaosBagResults: newChaosBagResults });
```
Comment (zzorba) [inline components/campaign/CampaignChaosBagView/index.tsx]: Actually, second problem.  You need to use the updateCampaign that comes out of props, not the action that's imported.  I've made this mistake SO many times myself.

File: actions/types.ts [hunk 20]
```diff
@@ -265,6 +277,7 @@ export interface Campaign {
   latestDeckIds?: number[]; // deprecated
   investigatorData: InvestigatorData;
   chaosBag: ChaosBag;
+  chaosBagResults: ChaosBagResults;
```
Comment (zzorba) [inline actions/types.ts]: So the legacy campaigns on devices means we need to make this optional
`chaosBagResults?: ChaosBagResults;`

Pretty sure the code is already setup to handle it being optional (with an `|| DEFAULT` in the connect statement)

File: components/campaign/actions.ts [hunk 21]
```diff
@@ -55,6 +55,7 @@ export function newCampaign(
   difficulty: CampaignDifficulty,
   deckIds: number[],
   chaosBag: ChaosBag,
+  chaosBagResults: ChaosBagResults,
```
Comment (zzorba) [inline components/campaign/actions.ts]: Was the places where `newCampaign` is called updated as well.

My guess is any 'new' campaign is going to always have an empty set of 'drawn' chaosBag tokens, so maybe we just leave this out and put a default down below in what gets dispatched?

Comment (zzorba) [reply components/campaign/actions.ts]: Yeah, I see the default is inserted below, maybe this `newCampaign` edit isn't needed anymore?
